### ISSUE TITLE
Is it possible to write to physical disk volumes?

### ISSUE THREAD

**Issue by rohitkiran77** [DiscUtils/DiscUtils#181]
Hi,

  I need to write to physical disk volumes, 
  i am looking to open physical disk , as (diskutil) ntfs file system in code.
  and then perform FileCopy like Utils/OSClone.. as i need to set attributes too.

 i am struggling to find a way to open physical volume and initializeFS ..
 when i tried, physical volume as  fileStreams, i got error where stream length is not set.

what is the way out here?

**Comment by LordMike** [comment]
You should be able to open filestreams to physical disks, and not have issues with lengths not being present.. (if that's the error ..)

In my other library, RawDiskLib, I've wrapped the necessary stuff to open streams to disks.. You could try that out.

**Comment by rohitkiran77** [comment]
Initilized  RawDisk and generated Stream from it, and used it in diskutils NTFS file system.
but when trying to write , new file

got error 

> System.UnauthorizedAccessException
  HResult=0x80070005
  Message=Access to the path is denied.
  Source=mscorlib
  StackTrace:
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.WriteCore(Byte[] buffer, Int32 offset, Int32 count)
   at System.IO.FileStream.Write(Byte[] array, Int32 offset, Int32 count)
   at RawDiskLib.RawDiskStream.Write(Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\RawDiskLib\RawDiskLib\RawDiskStream.cs:line 109
   at DiscUtils.Streams.SparseStream.SparseWrapperStream.Write(Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Streams\SparseStream.cs:line 304
   at DiscUtils.Streams.BlockCacheStream.Write(Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Streams\Block\BlockCacheStream.cs:line 396
   at DiscUtils.Ntfs.RawClusterStream.WriteClusters(Int64 startVcn, Int32 count, Byte[] buffer, Int32 offset) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\RawClusterStream.cs:line 344
   at DiscUtils.Ntfs.NonResidentAttributeBuffer.Write(Int64 pos, Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\NonResidentAttributeBuffer.cs:line 158
   at DiscUtils.Ntfs.NtfsAttributeBuffer.Write(Int64 pos, Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\NtfsAttributeBuffer.cs:line 162
   at DiscUtils.Streams.BufferStream.Write(Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Streams\Buffer\BufferStream.cs:line 174
   at DiscUtils.Streams.SparseStream.SparseWrapperStream.Write(Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Streams\SparseStream.cs:line 304
   at DiscUtils.Streams.BlockCacheStream.Write(Byte[] buffer, Int32 offset, Int32 count) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Streams\Block\BlockCacheStream.cs:line 396
   at DiscUtils.Ntfs.Bitmap.SetByte(Int64 index, Byte value) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\Bitmap.cs:line 218
   at DiscUtils.Ntfs.Bitmap.MarkPresent(Int64 index) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\Bitmap.cs:line 71
   at DiscUtils.Ntfs.Bitmap.AllocateFirstAvailable(Int64 minValue) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\Bitmap.cs:line 172
   at DiscUtils.Ntfs.MasterFileTable.AllocateRecord(FileRecordFlags flags, Boolean isMft) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\MasterFileTable.cs:line 283
   at DiscUtils.Ntfs.NtfsFileSystem.AllocateFile(FileRecordFlags flags) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\NtfsFileSystem.cs:line 2387
   at DiscUtils.Ntfs.File.CreateNew(INtfsContext context, FileRecordFlags flags, FileAttributeFlags dirFlags) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\File.cs:line 219
   at DiscUtils.Ntfs.File.CreateNew(INtfsContext context, FileAttributeFlags dirFlags) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\File.cs:line 214
   at DiscUtils.Ntfs.NtfsFileSystem.CreateNewFile(String path, NewFileOptions options) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\NtfsFileSystem.cs:line 2022
   at DiscUtils.Ntfs.NtfsFileSystem.OpenFile(String path, FileMode mode, FileAccess access, NewFileOptions options) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\NtfsFileSystem.cs:line 1751
   at DiscUtils.Ntfs.NtfsFileSystem.OpenFile(String path, FileMode mode, FileAccess access) in C:\Work\Hobby\DiscUtils\Library\DiscUtils.Ntfs\NtfsFileSystem.cs:line 665

>

**Comment by LordMike** [comment]
.. and you're sure to put ReadWrite as the permissions for the new handle?

**Comment by rohitkiran77** [comment]
the concerned code 
Where i open up the file , exception happens in Write API
`            string destFile = Path.Combine(destPath,Path.GetFileName(srcFile));
            using (Stream s = sourceNtfs.OpenFile(srcFile, FileMode.Open, FileAccess.Read))
            using (Stream d = destNtfs.OpenFile(destFile, FileMode.Create, FileAccess.Write))
            {
                d.SetLength(s.Length);
                int numRead = s.Read(_copyBuffer, 0, _copyBuffer.Length);
                while (numRead > 0)
                {
                    d.Write(_copyBuffer, 0, numRead);
                    numRead = s.Read(_copyBuffer, 0, _copyBuffer.Length);
                }
            }
`

Where i open up the device  (caller to above function)

`            DriveInfo cDrive = new DriveInfo("C");
            using (RawDisk pDrive = new RawDisk(cDrive, FileAccess.ReadWrite))
                {
                    using (NtfsFileSystem pNtfsFS = new NtfsFileSystem(pDrive.CreateDiskStream()))
                    {
                        using (FileStream virtualStream = new FileStream(ImageFilePath, FileMode.Open, FileAccess.Read, FileShare.Read))
                        {
                            using (NtfsFileSystem vNtfsFS = new NtfsFileSystem(virtualStream))
                            {
                                CopyFile(vNtfsFS, diskFile, pNtfsFS, @"\Personal\CSWorkBook\test\destfolder\");

                            }
                        }

                    }
                }`

**Comment by LordMike** [comment]
Note that in your example, `virtualStream` is opened with Read access - which might be ok, I'm not sure if it's source or destination in the code above. Ensure that the default for new NtfsFileSystem is a read-write implementation.

**Comment by rohitkiran77** [comment]
pNtfsFS  is the destination and  virtualStream is  source 

Adding signature of the function to to map the variables.

`private static void CopyFile(NtfsFileSystem sourceNtfs, string srcFile, NtfsFileSystem destNtfs, string destPath)
  `

**Comment by LordMike** [comment]
Can you confirm that this works fine with f.ex. a memory stream as destination?

Because then we know the issue is not DiscUtils, but the stream for the physical disk.

**Comment by rohitkiran77** [comment]
Instead of using memory stream , i wrote into  file stream initialized as NTFS ,  (basically a NTFS disk image as file) 
the write went through fine.. (created file with proper extension as expected)

**Comment by rohitkiran77** [comment]
One doubt i have is

>   Message=Access to the path is denied.
  Source=mscorlib
  StackTrace:
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)

Is this error because of not using filter driver to access NTFS hard drive.

**Comment by LordMike** [comment]
Then we have an issue with the raw disk handles. Oh well.

**Comment by EricZimmerman** [comment]
Doesn't Windows prevent writing to a raw volume?

**Comment by rohitkiran77** [comment]
> Doesn't Windows prevent writing to a raw volume?

Is there any way out? FSCTL_LOCK_VOLUME  and locking volume helps?

**Comment by LordMike** [comment]
So. In effect, yes. My current understanding is that Windows locks ranges of the disk (when mounted). So if we could trick windows to not mount something, it'd work.

What works:

1. Create a new volume in Windows, do not format it. DO assign it a drive letter (else it's just "not ready")
2. Identify its volume number (I used WinObj from SysInternals), it's likely the highest HarddiskVolume available.
3. In RawDisk/DiscUtils, open it and format as NTFS
4. Write a file to it in DU.

But. The moment I "rescan disks" in Disk Management, or open Explorer to all drives after step 3, Windows identifies the newly formatted drive as NTFS, and mounts it. From then on, writing anything to the drive is disallowed.

**Comment by rohitkiran77** [comment]
Thanks a lot LordMike  for the help.

**Comment by jonathanjsimon** [comment]
@LordMike I think I know what's going on here. I'm still testing my theory but I was able to write and entire disk image to disk without issue by jumping over the header at the top of the file. I went back and wrote that last. The problem is that as soon as the OS sees a mountable partition described in the partition table, it does exactly what you said it does, lock and mount. So if you don't have a partition table until the very end, there is nothing to mount.

In my case, I'm writing a disk image to a GPT disk, so I jumped over 16k at the top and went back to write it later. I think I borked an offset because it wasn't mountable afterwards but it didn't throw this exception either and it wrote the whole file to disk as opposed to almost none. I'll report back tomorrow if I get it to work.

**Comment by jonathanjsimon** [comment]
@LordMike Hah, I did see that comment about a day after I wrote my comment and was glad my idea was also though of my others. Made me feel more confident that it would work. It did, in fact, work and is now in production doing exactly what I said. The only other change I had to make was to read and write in larger chunks instead of byte by byte. I'm reading and writing in 5mb chunks; it's fast and reliable.
